PSPs/drafts/psp-6.md

```diff
@@ -0,0 +1,2903 @@
+
+#### 1.9.3.1. Parameter
+
+* `HASH` - (OPTIONAL) The block hash.
+
+#### 1.9.3.2. Response
+
+* `MAP` - (OPTIONAL)
+  * `"block"`: `MAP`
+    * `"extrinsics"`: `ARRAY`
+      * `HEX` - An extrinsic.
+    * `"header"`: `MAP`
+      * `"parentHash"`: `HASH` - The parent hash.
+      * `"number"`: `HEX` - The block number.
+      * `"stateRoot"`: `HASH` - The state trie merkle root.
+      * `"extrinsicsRoot"`: `HASH` - The merkle root of the extrinsics.
+      * `"digest"`: `MAP` - Chain-specific digest of data useful for light clients or auxiliary data.
+        * `"logs"`: `ARRAY` - A list of logs in the digest.
+          * `HEX` - Digest logs of opaque nature.
+  * `"justification"`: `HEX` - (OPTIONAL) The justification.
+
+#### 1.9.3.3. Example
+
+Request:
+
+```bash
+curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "chain_getBlock", "params":[]}' http://localhost:9933
+```
+
+Response:
+
+```json
+{
+    "jsonrpc": "2.0",
+    "result": {
+        "block": {
+            "extrinsics": [
+                "0x280402000bb0019c8b7001",
+                "0x1c0409006a4b4b00",
+                "0x1004140000"
+            ],
+            "header": {
+                "digest": {
+                    "logs": [
+                        "0x06424142453402610000007d7fb90f00000000",
+                        "0x0542414245010166b01406e966b6604ea996cda5355c5f24fffee6cb2082802105f136b456c9412855a0cce4058f08ccf9ce822cd94a6ae786037a8566310765a043a09514488a"
+                    ]
+                },
+                "extrinsicsRoot": "0x1ce1236c8262cead39a2a758fc4cdba0b918f309e5a8572c288c5ba2c16aa7a7",
+                "number": "0x12d2dd",
+                "parentHash": "0x78b7af6942c220ec33945f228f1e17383403561f8d7bbc02da2c79650f22fdbe",
+                "stateRoot": "0xa6387842acac9315ad4424828f571c348a40bdfb9f9b4cb4951bcb76e43dacfb"
+            }
+        },
+        "justification": null
+    },
+    "id": 1
+}
+```
+
+### 1.9.4. chain_getBlockHash
+
+Get hash of the 'n-th' block in the canon chain. If no parameters are provided, the latest block hash gets returned.
+
+Alias: `chain_getHead`
+
+#### 1.9.4.1. Parameter
+
+One of the following types is allowed:
+
+* `HEX` - (OPTIONAL) The value indicating the 'n-th' block in the chain.
+* `U32` - (OPTIONAL) The value indicating the 'n-th' block in the chain.
+* `ARRAY` - (OPTIONAL) Multiple values, where each value can be one of the following types:
+  * `HEX` - The value indicating the 'n-th' block in the chain.
+  * `U32` - The value indicating the 'n-th' block in the chain.
+
+#### 1.9.4.2. Response
+
+One of the following types gets returned, depending on the provided parameters:
+
+* `HASH` - (OPTIONAL) The block hash.
+* `ARRAY`
+  * `HASH` - (OPTIONAL) A block hash
+
+#### 1.9.4.3. Example
+
+Request:
+
+```bash
+curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "chain_getBlockHash", "params":[50, "0x64", [200]]}' http://localhost:9933
```

Comment:
One of seem to suggest just one of them, here the example uses all three at once. Is that as intended?

Reply:
The parameters on this method work somewhat interestingly. You can specify one or multiple block numbers, either as integers or hex string, and can even include embedded arrays. But I guess this is not really "good design", so I'll adjust this (backwards compatible with substrate).